### PR TITLE
Inverting TypeOf should support 'TypeOf x IsNot...'

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/51359

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/InvertIf/InvertIfTests.Elseless.cs [hunk 1]
```diff
@@ -837,5 +838,110 @@ public string Prop
...
+        {
+            return 2;
+        }
+    }
+}",
+@"class C
+{
+    int M()
+    {
+        if (c is not object)
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/InvertIf/InvertIfTests.Elseless.cs:900]: 💭 This really should be:

```suggestion
        if (c is not null)
```

File: src/EditorFeatures/CSharpTest/InvertIf/InvertIfTests.Elseless.cs [hunk 2]
```diff
@@ -837,5 +838,110 @@ public string Prop
...
+        {
+            return 2;
+        }
+    }
+}",
+@"class C
+{
+    int M()
+    {
+        if (!(c is object))
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/InvertIf/InvertIfTests.Elseless.cs:865]: 💭 Even in C# 8 we can use `if (c is null)` here

**Comment by CyrusNajmabadi** [comment]
@sharwell  i have a followup PR to our other features to improve those scenarios :)

**Comment by CyrusNajmabadi** [comment]
@sharwell this is ready for rereview.

File: src/EditorFeatures/CSharpTest/InvertIf/InvertIfTests.Elseless.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/InvertIf/InvertIfTests.Elseless.cs]: yup.  we do this now.

File: src/EditorFeatures/CSharpTest/InvertIf/InvertIfTests.Elseless.cs [hunk 3]
```diff
@@ -837,5 +838,110 @@ public string Prop
...
+        {
+            return 2;
+        }
+    }
+}",
+@"class C
+{
+    int M()
+    {
+        if (c is object)
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/InvertIf/InvertIfTests.Elseless.cs]: done.

File: src/EditorFeatures/CSharpTest/InvertIf/InvertIfTests.Elseless.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/InvertIf/InvertIfTests.Elseless.cs]: done.

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs [hunk 4]
```diff
@@ -412,21 +442,46 @@ private static IOperation RemoveImplicitConversion(IOperation operation)
...
+            SyntaxNode pattern,
+            SyntaxGenerator generator,
+            SyntaxGeneratorInternal generatorInternal,
+            ISyntaxFacts syntaxFacts)
+        {
+            Contract.ThrowIfFalse(syntaxFacts.IsNotPattern(pattern));
+
+            syntaxFacts.GetPartsOfUnaryPattern(pattern, out var opToken, out var subPattern);
+
+            // If we started with `not object`, instead of converting to `object`, directly convert to `not null`
```
Comment (davidwengier) [inline src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:470]: I can't decide how I feel about this. `is not object` has been valid as long as `is null`, right? So someone writing it presumably did so deliberately, so isn't the negation of it `is object`?

I guess we'll hear about it if anyone doesn't like it :)

**Review by davidwengier — APPROVED** [review]

File: src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Extensions/SyntaxGeneratorExtensions_Negate.cs:470]: > I can't decide how I feel about this. `is not object` has been valid as long as `is null`, right?

Nope!  Patterns were added in C#7.   `not` patterns though only came in C#9.

**Pull request merged**